syntheseus/search/algorithms/best_first/base.py

```diff
@@ -2,17 +2,82 @@
 
 import abc
 import heapq
+import itertools
 import logging
 import math
 from collections.abc import Sequence
-from typing import Generic
+from dataclasses import dataclass, field
+from typing import Any, Generic
 
 from syntheseus.search.algorithms.base import GraphType, SearchAlgorithm
 from syntheseus.search.graph.node import BaseGraphNode
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass(order=True)
+class PriorityQueueItem:
+    priority: float
+    tie_breaker: int
+    item: Any = field(compare=False)
+
+
+class PriorityQueue:
+    """
+
+    Simple priority queue implementation which supports removing elements
+    and changing their priority.
+
+    Implementation based on heapq and guidance from:
+    https://docs.python.org/3/library/heapq.html#priority-queue-implementation-notes
+
+    In particular, this means items are never really removed from the queue:
+    they are just marked as invalid. This is why the queue length is not
+    necessarily the same as the number of unique items in the queue.
+    It also assumes that items are unique and hashable.
+    """
+
+    def __init__(self):
+        self._queue = []
+        self._entry_finder = {}
+        self._counter = itertools.count()  # to break ties in priority
+        self._REMOVED = None
```

Comment:
Is `self._REMOVED` always just an alias for `None`?

Reply:
At the moment yes. Would you recommend having the option to override or customize this?

Reply:
What is the rationale for customizing this? Is it that currently it's not possible to put in an item that is equal to `None` because the `if entry.item is not self._REMOVED` logic will get confused, and setting `_REMOVED` to something else would allow that?

Reply:
Yes, that would be the rationale to customize it. But at the moment I didn't intend to put `None` into the priority queue so I thought it was fine to hard code it.

I was just suggesting the replacement because I didn't know why you made the comment.

Reply:
If we don't expect to put `None` items then my hunch would be to just directly use `None` for simplicity (and skip adding an "intermediate" in the form of `self._REMOVED`). Or we could also replace `if entry.item is not self._REMOVED` with `if entry.item in self._entry_finder` if we don't mind an additional dictionary lookup (which would remove the need to ever compare `entry.item` to the special removed value and we could use `None` even if some items are `None` to being with).

Reply:
This is a fair suggestion. I changed `self._REMOVED` to a hard-coded `None`. However, I don't think adding `None` to the queue will work properly with your suggested code change. Imagine in pseudo-code:

```
queue = PQ([])
queue.push(A, 0)  # queue is [(0, A)]
queue.change_priority(A, 1)  # queue will now be [(0, None), (1, A)]
queue.push(None, 2)  # queue will now be [(0, None), (1, A), (2, None)]
queue.pop()  # will return `(0, None)` which is incorrect
```

Instead I changed `if entry.item is not self._REMOVED` to `self._entry_finder.get(entry.item) == entry`

Does this seem like a good change?